packages/website/ts/components/forms/subscribe_form.tsx

```diff
@@ -0,0 +1,121 @@
+                <Container marginBottom="15px">
+                    <Text fontFamily="Roboto Mono" fontColor={colors.grey} center={true}>
+                        Subscribe to our newsletter for 0x relayer and dApp updates
+                    </Text>
+                </Container>
+                <form onSubmit={this._handleFormSubmitAsync.bind(this)}>
+                    <Container className="flex flex-wrap justify-center items-center">
+                        <Container marginTop="15px">
+                            <Input
+                                placeholder="you@email.com"
+                                value={this.state.emailText}
+                                fontColor={colors.white}
+                                fontSize={formFontSize}
+                                backgroundColor={colors.projectsGrey}
+                                width="275px"
+                                onChange={this._handleEmailInputChange.bind(this)}
+                            />
+                        </Container>
+                        <Container marginLeft="15px" marginTop="15px">
+                            <Button
+                                type="submit"
+                                backgroundColor={colors.darkButtonGrey}
+                                fontColor={colors.white}
+                                fontSize={formFontSize}
+                            >
+                                Subscribe
+                            </Button>
+                        </Container>
+                    </Container>
+                </form>
+                {this._renderMessage()}
+            </Container>
+        );
+    }
+    private _renderMessage(): React.ReactNode {
+        let message = null;
+        switch (this.state.status) {
+            case SubscribeFormStatus.Error:
+                message = 'Sorry, something went wrong. Try again later.';
+                break;
+            case SubscribeFormStatus.Loading:
+                message = 'One second...';
+                break;
+            case SubscribeFormStatus.Success:
+                message = `Thanks! ${this.state.lastSubmittedEmail} is now on the mailing list.`;
+                break;
+            case SubscribeFormStatus.None:
+                break;
+        }
+        return (
+            <Container isHidden={!message} marginTop="30px">
+                <Text center={true} fontFamily="Roboto Mono" fontColor={colors.grey}>
+                    {message || 'spacer text'}
+                </Text>
+            </Container>
+        );
+    }
+
+    private _handleEmailInputChange(event: React.ChangeEvent<HTMLInputElement>): void {
+        this.setState({ emailText: event.target.value });
+    }
+    private async _handleFormSubmitAsync(event: React.FormEvent<HTMLInputElement>): Promise<void> {
+        event.preventDefault();
+        if (!this.state.emailText) {
```

Comment:
Let's use `_.isEmpty` or `_.isUndefined` here. Otherwise this conditional feels too loose.

Reply:
I've noticed our convention of using these utils in the codebase, could you briefly walk me through the advantages? To me it seems like "loose" is a good quality to have in this case, as we are catching `null | undefined | ''` and the rest is caught by typescript. I see how it is more explicit, but don't see other advantages. I personally like relying on javascript falsyness because it is often exactly what I mean, but maybe I have bad habits :P

Reply:
Sure. First off, we actually avoid using `null` values in the codebase. Instead we always use `undefined`. This came out of a longer discussion with @LogvinovLeon where we reasoned that JS returns `undefined` in cases where a property or value does not exist. JS never returns `null`. 

The main question we had was: "Is there a good reason to distinguish between `undefined` and `null` values? Do they provide some semantic meaning to the developer?". The official answer is that `undefined` refers to a value that has been declared but not assigned. `null` must be intentionally assigned and signifies the value has been intentionally assigned to an empty or non-existent value. Although is difference could make sense in some niche cases, most of the time, we simply care if the value is non-existent or existent. 

To simplify the codebase and to enforce stricter type checking, we always use `undefined` and check with `_.isUndefined` (and `_.isEmpty` for string values that can be empty). Being stricter in the conditional checking means that we won't accidentally have a "falsy" value such as `0`, `false`, `NaN` or `''` trigger the conditional, then what we really wanted to check for was whether the value was defined or not. We've found that we find bugs much sooner by having this stricter policy.

Does this make sense? Like always, we are open to changing any code convention if an alternative proposition makes more rational sense. The only downside of our approach that we've encountered so far is having to wrap some public interfaces of dependencies that currently return `null` rather then `undefined`. We also make an exception for JSX returning methods since `null` is special-cased in React to avoid rendering elements to the DOM.

Reply:
That makes sense, although this specific example may not be the best show-case for it. If I use `_.isEmpty` instead of just checking for falsyness, I actually open myself up to bugs as I have to assume that `emailText` is always a non-null non-undefined string. Typescript can guarantee me that it is a string, but with the current compiler settings (in `website`) it cannot guarantee me that it is not `null` or `undefined`. You mention that we don't use `null`, so that narrows it down to `undefined`. So I guess our convention in this case would be to use both `_.isUndefined` and `_.isEmpty`. However, as you mentioned there are some cases in which `null` does sneak in (we have to allow it for JSX and 3rd party libraries tend to use it), so we are still making an assumption and hoping for the best.

So I guess I disagree :P. But this is code is a closed system and I think following convention (and shipping) trumps some argument about how it could theoretically be incorrect. Maybe we can revisit this in the future, going to use `isUndefined` and `isEmpty` for now.